runtime/compiler/runtime/IProfiler.cpp

```diff
@@ -4853,67 +4901,309 @@ void TR_IProfiler::dumpIPBCDataCallGraph(J9VMThread* vmThread)
       fprintf(stderr, "Cannot allocate memory. Bailing out.\n");
       return;
       }
+   traverseIProfilerTableAndCollectEntries(&aggregationHT, vmThread, true/*collectOnlyCallGraphEntries*/);
+   aggregationHT.sortByNameAndPrint();
 
-   J9JavaVM *javaVM = vmThread->javaVM;
-   J9InternalVMFunctions *vmFunctions = javaVM->internalVMFunctions;
-   TR_J9VMBase * fe = TR_J9VMBase::get(javaVM->jitConfig, vmThread);
+   fprintf(stderr, "Finished dumping info\n");
+   }
 
-   TR::VMAccessCriticalSection dumpCallGraph(fe);
+uintptr_t
+TR_IProfiler::createBalancedBST(TR_IPBytecodeHashTableEntry **ipEntries, int32_t low, int32_t high, uintptr_t memChunk, TR_J9SharedCache *sharedCache)
+   {
+   if (high < low)
+      return 0;
 
-   fprintf(stderr, "Aggregating per method ...\n");
-   for (int32_t bucket = 0; bucket < TR::Options::_iProfilerBcHashTableSize; bucket++)
+   TR_IPBCDataStorageHeader * storage = (TR_IPBCDataStorageHeader *) memChunk;
+   int32_t middle = (high+low)/2;
+   TR_IPBytecodeHashTableEntry *entry = ipEntries[middle];
+   uint32_t bytes = entry->getBytesFootprint();
+   entry->createPersistentCopy(sharedCache, storage, _compInfo->getPersistentInfo());
+
+   uintptr_t leftChild = createBalancedBST(ipEntries, low, middle - 1, memChunk + bytes, sharedCache);
+   if (leftChild)
       {
-      //fprintf(stderr, "Looking at bucket %d\n", bucket);
-      for (TR_IPBytecodeHashTableEntry *entry = _bcHashTable[bucket]; entry; entry = entry->getNext())
+      TR_ASSERT(bytes < 1 << 8, "Error storing iprofile information: left child too far away"); // current size of left child
+      storage->left = bytes;
+      }
+
+   uintptr_t rightChild = createBalancedBST(ipEntries, middle + 1, high, memChunk + bytes + leftChild, sharedCache);
+   if (rightChild)
+      {
+      TR_ASSERT(bytes + leftChild < 1 << 16, "Error storing iprofile information: right child too far away"); // current size of right child
+      storage->right = bytes+leftChild;
+      }
+
+   return bytes + leftChild + rightChild;
+   }
+
+// Persist all IProfiler entries into the SCC.
+// This is done by aggregating all IProfiler entries per
+// ROMMethod with the help of a TR_AggregationHT table.
+// Then, for each ROMMethod with IP info we store the
+// entries arranged as a BST (for fast retrieval later).
+void
+TR_IProfiler::persistAllEntries()
+   {
+   J9JavaVM * javaVM = _compInfo->getJITConfig()->javaVM;
+   J9VMThread *vmThread = javaVM->internalVMFunctions->currentVMThread(javaVM);
+   TR_J9VMBase * fe = TR_J9VMBase::get(_compInfo->getJITConfig(), vmThread, TR_J9VMBase::AOT_VM);
+   TR_J9SharedCache *sharedCache = fe->sharedCache();
+   static bool SCfull = false;
+
+   if (!(TR::Options::sharedClassCache() && sharedCache))
+      return;
+
+   //int locked = releaseAllEntries();
```

Comment:
Does this commented line need to be included in this PR?

Reply:
I will delete it.